# 9 · Vector search — similarity, in the database you already have

Exact cosine similarity over nodes **and** edges, on plain `real[]` columns.
No pgvector, no extension, no approximate index — and because the search is
exact rather than an index probe, filtering by metadata makes it *cheaper*
rather than fighting a recall budget.

Three things this notebook shows in order: similarity as a **search**
(`vector_search`), similarity as part of a **traversal** (`Start(near=)`,
`Hop(near=)`, `Hop(via_near=)`), and the two places the design deliberately
**refuses** — a vector inside a filter, and a vector inside an LLM tool call.

Real embeddings are 384 to 3072 opaque floats, which makes for documentation
nobody can check. So this notebook uses **named axes** instead — interests,
roles, and what kind of tie an edge is. Every ranking below is verifiable by
eye against the numbers in the next cell, and hopai neither knows nor cares
that they mean anything: it stores whatever floats your embedding model
produces.

In [1]:
from demo_graph import arrows, connect, names, seed
from hopai import Boost, Hop, Near, Start, Vector

graph = connect("nb_09_vectors")
ids = seed(graph)

#: The axes a real embedding model would learn. Named here so a ranking
#: can be checked by reading, rather than believed.
INTERESTS = ["databases", "design", "finance", "outdoors"]
SUMMARY = {
    "Alice": [0.90, 0.30, 0.10, 0.20],   # databases
    "Bob":   [0.20, 0.90, 0.10, 0.20],   # design
    "Carol": [0.10, 0.10, 0.95, 0.20],   # finance
    "Dave":  [0.10, 0.20, 0.10, 0.95],   # outdoors
    "Erin":  [0.60, 0.70, 0.10, 0.10],   # databases + design
}
ROLES = ["ic", "manager", "founder"]
TITLE = {                                # Erin has none -- the missing case
    "Alice": [0.95, 0.10, 0.05],
    "Bob":   [0.20, 0.90, 0.10],
    "Carol": [0.10, 0.40, 0.90],
    "Dave":  [0.30, 0.80, 0.10],
}
TIES = ["social", "professional"]
TIE = {"friend": [0.95, 0.15], "works_at": [0.10, 0.95]}

print(f"{len(SUMMARY)} people; axes: {INTERESTS} / {ROLES} / {TIES}")

5 people; axes: ['databases', 'design', 'finance', 'outdoors'] / ['ic', 'manager', 'founder'] / ['social', 'professional']


## Declaring the fields

A vector field is a **named** column, declared per target and sized once.
Declaring is instant; the `ALTER TABLE` it implies is a separate, explicit
call, because a migration should be a conscious moment rather than a side
effect of an import. `vector_ddl()` shows exactly what that call would run:

In [2]:
graph.define_vectors(nodes=[Vector("summary", 4), Vector("title", 3)],
                     edges=[Vector("relation", 2)])

for statement in graph.vector_ddl():
    print(statement, "\n")

ALTER TABLE "nodes" ADD COLUMN IF NOT EXISTS "vec_summary" real[] 

ALTER TABLE "nodes" ALTER COLUMN "vec_summary" SET STORAGE EXTERNAL 

ALTER TABLE "nodes" ADD CONSTRAINT "ck_vec_dims_default_nodes_summary" CHECK (graph_id != 'default' OR vec_summary IS NULL OR array_ndims(vec_summary) = 1 AND array_length(vec_summary, 1) = 4) 

ALTER TABLE "nodes" ADD COLUMN IF NOT EXISTS "vec_title" real[] 

ALTER TABLE "nodes" ALTER COLUMN "vec_title" SET STORAGE EXTERNAL 

ALTER TABLE "nodes" ADD CONSTRAINT "ck_vec_dims_default_nodes_title" CHECK (graph_id != 'default' OR vec_title IS NULL OR array_ndims(vec_title) = 1 AND array_length(vec_title, 1) = 3) 

ALTER TABLE "edges" ADD COLUMN IF NOT EXISTS "vec_relation" real[] 

ALTER TABLE "edges" ALTER COLUMN "vec_relation" SET STORAGE EXTERNAL 

ALTER TABLE "edges" ADD CONSTRAINT "ck_vec_dims_default_edges_relation" CHECK (graph_id != 'default' OR vec_relation IS NULL OR array_ndims(vec_relation) = 1 AND array_length(vec_relation, 1) = 2) 



Three statements per field, and each is a decision:

- **`real[]`, beside `properties`, never inside it.** A 1536-float array in
  JSONB would bloat the GIN index and ride along in every traversal result.
- **`SET STORAGE EXTERNAL`.** Float noise does not compress, so TOAST
  compression spends CPU to save nothing.
- **A per-graph `CHECK`, not a typed `vector(d)`.** The dimension is guarded
  by a constraint scoped to `graph_id`, which is what lets two graphs share
  one column at *different* dimensionality — something a typed column cannot
  express. Later in this notebook, they do.

`migrate_vectors()` applies it. It is idempotent — safe next to
`create_schema()` in a start-up path — and it **refuses drift** rather than
papering over it: a column of the wrong type, or an existing constraint
declaring different dimensions, names the fix instead of quietly serving two
incompatible definitions.

In [3]:
print(graph.migrate_vectors())
print(graph.migrate_vectors(), "<- again: the same fields ENSURED, no DDL applied")

['nodes.vec_summary', 'nodes.vec_title', 'edges.vec_relation']
['nodes.vec_summary', 'nodes.vec_title', 'edges.vec_relation'] <- again: the same fields ENSURED, no DDL applied


## Storing

`set_vectors()` is the **only** write path for a vector. `add_nodes()` and
`merge_nodes()` do not accept one, and that is deliberate: an embedding is
produced by a model call, on its own latency and failure budget, so tying it
to row creation would make every ingest wait on it. The whole call is one
transaction.

In [4]:
rows = [{"id": ids[name], "summary": vector} for name, vector in SUMMARY.items()]
for row in rows:                                    # one row per id, both fields
    name = next(n for n, i in ids.items() if i == row["id"])
    if name in TITLE:
        row["title"] = TITLE[name]

print(graph.set_vectors(nodes=rows), "node rows updated")

5 node rows updated


Acme and Globex get no `summary` and Erin no `title` — those absences are the
missing-vector cases this notebook turns on later.

Edges are embedded the same way, but their ids need looking up first: a
traversal returns a *subgraph*, and an edge there is reported by its
endpoints and properties rather than by an id. The edge rows are ordinary
table rows, so one select is enough.

In [5]:
from sqlalchemy import select

edges_tbl = graph.edges_tbl
with graph.engine.connect() as connection:
    edge_kinds = dict(connection.execute(
        select(edges_tbl.c.id, edges_tbl.c.properties["kind"].astext)
        .where(edges_tbl.c.graph_id == graph.graph)).all())

print(graph.set_vectors(edges=[{"id": edge_id, "relation": TIE[kind]}
                               for edge_id, kind in edge_kinds.items()]), "edges embedded")

9 edges embedded


Reading a vector back is an explicit call, by id. It is **not** part of any
traversal or search result — six kilobytes of floats has no business in every
subgraph, and nothing downstream ever needs them:

In [6]:
print(graph.get_vectors(node_ids=[ids["Alice"]]))
print()
print(graph.traverse(Start(where={"name": "Alice"})).nodes, "<- no vector here")

{'nodes': {'1': {'summary': [0.9, 0.3, 0.1, 0.2], 'title': [0.95, 0.1, 0.05]}}, 'edges': {}}

[{'id': '1', 'properties': {'age': 34, 'city': 'Berlin', 'name': 'Alice', 'type': 'person', 'email': 'alice@example.com', 'active': True}}] <- no vector here


## Searching

`vector_search()` ranks by cosine similarity and returns the hits most similar
first, each with its score. Ask for the people most interested in databases —
axis 0:

In [7]:
databases = [1.0, 0.0, 0.0, 0.0]

for hit in graph.vector_search(Near("summary", databases), k=3):
    print(f"{hit['properties']['name']:<6} {hit['similarity']:.3f}")

Alice  0.923
Erin   0.643
Bob    0.211


Alice first, Erin second, Bob third — reading the vectors above, that is the
order the numbers demand, and the scores are real cosines you can recompute.

Two things are already visible in that result. Cosine ignores magnitude, so
only the *direction* of a vector matters (which is why every embedding API
ships unit-normalized vectors and why dot and euclidean would rank the same).
And Acme and Globex are absent even at `k=10`: a row with no vector for the
field has no similarity, so it is not ranked. Missing is missing, not zero.

In [8]:
everyone = graph.vector_search(Near("summary", databases), k=10)
print([hit["properties"]["name"] for hit in everyone])
assert "Acme" not in [hit["properties"]["name"] for hit in everyone]

['Alice', 'Erin', 'Bob', 'Carol', 'Dave']


### Filtering is free

`where=` is the same filter language as traversal, and it is applied **before**
ranking. Because the scan is exact rather than an approximate index probe, a
selective filter makes the search cheaper — there is no recall cliff to trade
against, which is the failure mode filtered ANN search is famous for:

In [9]:
for hit in graph.vector_search(Near("summary", databases), k=3, where={"city": "Berlin"}):
    print(f"{hit['properties']['name']:<6} {hit['similarity']:.3f}")

Alice  0.923
Bob    0.211


### Several fields at once

Multivector search: each `Near` contributes its own cosine, scaled by its
weight, and the sum is what ranks. Asking for a database-minded **manager**
combines the two fields — and Erin, who has no `title` vector, drops out:

In [10]:
manager = [0.0, 1.0, 0.0]

for hit in graph.vector_search(Near("summary", databases, weight=0.5),
                               Near("title", manager, weight=0.5), k=5):
    print(f"{hit['properties']['name']:<6} {hit['similarity']:.3f}")

Bob    0.591
Dave   0.516
Alice  0.514
Carol  0.253


Bob wins: a manager, and more database-minded than Carol or Dave. His score is
half his interest similarity (`0.211`, from the first search above) plus half
his role similarity (`0.970`) — `0.591`, checkable by hand.

`missing=` decides what a row without this field's vector does. The default
`"exclude"` drops it — that is why Erin is not above. `"zero"` scores her 0 on
`title` and lets `summary` carry the row, which is what you want when a field
is genuinely optional rather than required:

In [11]:
for hit in graph.vector_search(Near("summary", databases, weight=0.5),
                               Near("title", manager, weight=0.5, missing="zero"), k=5):
    print(f"{hit['properties']['name']:<6} {hit['similarity']:.3f}")

Bob    0.591
Dave   0.516
Alice  0.514
Erin   0.322
Carol  0.253


A similarity is also missing — the same NULL, treated the same way — when the
stored vector is all zeros (no direction, so cosine is undefined) or **not the
declared length**. That last guard is not defensive noise: PostgreSQL's
`unnest(a, b)` pads the shorter array with NULLs, so a mis-sized vector would
otherwise score a confident cosine over whatever prefix the two share.

`min_similarity` is a floor on *one field's* cosine, applied before `k`:

In [12]:
for hit in graph.vector_search(Near("summary", databases, min_similarity=0.5), k=10):
    print(f"{hit['properties']['name']:<6} {hit['similarity']:.3f}")

Alice  0.923
Erin   0.643


### Hybrid ranking, and the one thing hopai will not guess

`Boost` adds a numeric property to the score — recency, popularity, an
editorial weight. It is a nudge, not a filter: a row where the property is
absent, null or non-numeric contributes `default` (0.0) instead of dropping
out. Use `where=` to filter.

And here is the rule the library refuses to guess for you. A cosine lives in
`[-1, 1]`. Boost a raw property that does not, and it does not *tilt* the
ranking, it **replaces** it — watch `age` bury the similarity entirely:

In [13]:
for hit in graph.vector_search(Near("summary", databases), boost=Boost("age", 0.2), k=5):
    print(f"{hit['properties']['name']:<6} {hit['similarity']:.3f}")

Dave   10.502
Bob    8.411
Alice  7.723
Carol  5.902
Erin   5.243


Dave — tied for *least* database-minded in the graph — is first, because he is
the oldest. hopai will not normalize on your behalf: doing so would
silently change which neighbours come back, and you would have no way to see
it happened.

Store the property already scaled to roughly `[0, 1]`, or scale it in SQL with
the callable form — the same escape hatch the filter DSL has:

In [14]:
from sqlalchemy import Float, cast

scaled = Boost(lambda properties: cast(properties["age"].astext, Float) / 100, weight=0.3)

print("no boost:  ", [(h["properties"]["name"], round(h["similarity"], 3))
                      for h in graph.vector_search(Near("summary", databases), k=4)])
print("with boost:", [(h["properties"]["name"], round(h["similarity"], 3))
                      for h in graph.vector_search(Near("summary", databases),
                                                   boost=scaled, k=4)])

no boost:   [('Alice', 0.923), ('Erin', 0.643), ('Bob', 0.211), ('Carol', 0.102)]


with boost: [('Alice', 1.025), ('Erin', 0.712), ('Bob', 0.334), ('Dave', 0.258)]


Two things in that output are worth reading twice. At `k=4` the boost did not
merely reorder — Dave replaced Carol, so **a boost with a limit decides
membership**, not just ranking. And Alice's score is now `1.025`: a boosted
`similarity` is the combined score, no longer a cosine, and can exceed 1.

## Similarity meets traversal

Three places, one per question you might be asking.

**`Start(near=, keep=)`** — start the walk from the nearest nodes. Compare
seeding from every person against seeding from the single most
database-minded one:

In [15]:
friends = Hop(via={"kind": "friend"}, hops=(1, 1))

everyone = graph.traverse(Start(where={"type": "person"}), friends)
nearest = graph.traverse(Start(near=Near("summary", databases), keep=1), friends)

print("from all 5 people:", names(everyone), arrows(everyone))
print("from the nearest: ", names(nearest), arrows(nearest))

from all 5 people: ['Alice', 'Bob', 'Carol', 'Dave', 'Erin'] ['Alice -friend-> Bob', 'Alice -friend-> Carol', 'Bob -friend-> Dave', 'Carol -friend-> Dave', 'Dave -friend-> Erin', 'Erin -friend-> Alice']
from the nearest:  ['Alice', 'Bob', 'Carol'] ['Alice -friend-> Bob', 'Alice -friend-> Carol']


With no hops at all, `Start(near=…, keep=N)` selects exactly what
`vector_search(…, k=N)` selects, minus the scores — which is why `near=` earns
its place on `Start` rather than being redundant: a traversal cannot be seeded
from a list of ids.

In [16]:
seeded = names(graph.traverse(Start(near=Near("summary", databases), keep=2)))
searched = sorted(h["properties"]["name"]
                  for h in graph.vector_search(Near("summary", databases), k=2))
print(seeded, searched)
assert seeded == searched

['Alice', 'Erin'] ['Alice', 'Erin']


**`Hop(near=, keep=)`** — walk normally, then keep only the nearest of what
the hop reached. Everyone within three friend-hops of Alice is everyone else;
keep the single most finance-minded and you get Carol — and the reported edges
shrink to the path that survived, because a traversal reports the edges it
really walked rather than every edge it considered:

In [17]:
result = graph.traverse(
    Start(where={"name": "Alice"}),
    Hop(via={"kind": "friend"}, hops=(1, 3), near=Near("summary", [0, 0, 1.0, 0]), keep=1),
)
print(names(result), arrows(result))

['Alice', 'Carol'] ['Alice -friend-> Carol']


**`Hop(via_near=, via_keep=)`** — rank the **edges** instead. This is a beam
*per source node*, not a global truncation, because "the nearest edges" only
means something relative to where you are standing. Bob and Carol each have a
`friend` edge and a `works_at` edge; ask for the most professional one out of
each and both keep their own:

In [18]:
from hopai import OR

both = Start(where=OR({"name": "Bob"}, {"name": "Carol"}))
for label, query in (("professional", [0.0, 1.0]), ("social", [1.0, 0.0])):
    result = graph.traverse(both, Hop(via_near=Near("relation", query), via_keep=1))
    print(f"{label:<13}", arrows(result))

professional  ['Bob -works_at-> Acme', 'Carol -works_at-> Acme']


social        ['Bob -friend-> Dave', 'Carol -friend-> Dave']


Two edges come back for `via_keep=1`, not one — that is the per-anchor beam
doing its job. The cycle guard sits *inside* the beam, so a beam never spends
its slots on edges leading back into the path it already walked.

One caveat worth stating plainly: **a traversal returns a subgraph, not a
ranking.** The scores and their order do not survive into the result. When
you need them, use `vector_search()`.

## Several queries in one round trip

A retrieval question expanded into sub-queries is one statement, not one per
query:

In [19]:
outdoors = [0.0, 0.0, 0.0, 1.0]

for batch in graph.vector_search_many([Near("summary", databases),
                                       Near("summary", outdoors)], k=2):
    print([(h["properties"]["name"], round(h["similarity"], 3)) for h in batch])

[('Alice', 0.923), ('Erin', 0.643)]
[('Dave', 0.968), ('Bob', 0.211)]


This buys **round trips, not arithmetic** — every query still scores every
candidate, measured at 1.08× against a loop on a local database. The saving is
`N-1` network round trips, which is where the cost actually is when Postgres
is not on localhost.

## One column, two graphs, two dimensionalities

The dimension is a `CHECK` scoped to `graph_id` rather than the column's type,
so a second graph in the same tables can size the same field differently — for
a cheaper embedding model, say. A typed `vector(4)` column could not express
this at all:

In [20]:
small = graph.in_graph("small-model")          # a fresh handle: no schema, no vectors
small.define_vectors(nodes=[Vector("summary", 2)])
small.migrate_vectors()

print(graph.vector_ddl()[2], "\n")
print(small.vector_ddl()[2])

ALTER TABLE "nodes" ADD CONSTRAINT "ck_vec_dims_default_nodes_summary" CHECK (graph_id != 'default' OR vec_summary IS NULL OR array_ndims(vec_summary) = 1 AND array_length(vec_summary, 1) = 4) 

ALTER TABLE "nodes" ADD CONSTRAINT "ck_vec_dims_a6796ac4e7_nodes_summary" CHECK (graph_id != 'small-model' OR vec_summary IS NULL OR array_ndims(vec_summary) = 1 AND array_length(vec_summary, 1) = 2)


Both constraints are live on the one `vec_summary` column, each guarding its
own graph. Note that the second is named with a **hash of the graph name**
rather than the name itself: PostgreSQL truncates identifiers at 63
characters, and truncating base and suffix independently once let two graphs
land on the same constraint name — silently disabling one graph's enforcement
and letting its `drop_vectors()` remove the other's.

The library refuses a wrong-sized vector before it ever reaches those
constraints, naming the declared size:

In [21]:
try:
    graph.set_vectors(nodes=[{"id": ids["Alice"], "summary": [1.0, 0.0]}])
except ValueError as exc:
    print(f"ValueError: {exc}")

ValueError: nodes '1': vector for 'summary' has 2 dimensions, the field is defined with 4


## Text in, vectors out

Every cell so far handed hopai floats. It does not have to be that way: give a
field the embedding client you already have and hopai calls it for you. Any
OpenAI, Cohere, Voyage or Google client works, as does a `SentenceTransformer`,
anything with LangChain's `embed_documents`/`embed_query`, LlamaIndex's
`get_text_embedding_batch`/`get_query_embedding`, or a plain function.

hopai imports **none** of those packages — clients are recognized by module name
and attribute shape, so `hopai[openai]` is a convenience alias for "hopai plus
the client you were installing anyway", never a coupling. Which is what lets
this notebook use a two-line model instead:

In [22]:
#: A stand-in for a real provider: one axis per city, scored by whether
#: the city is named. Deterministic and readable, which a real embedding
#: is not -- so every ranking below can be checked rather than believed.
CITIES = ["Berlin", "Lisbon"]


def city_model(texts):
    return [[float(city in text) for city in CITIES] for text in texts]


graph.define_vectors(
    nodes=[Vector("summary", 4), Vector("title", 3),
           Vector("place", 2, source="city", embed=city_model)],
    edges=[Vector("relation", 2)],
)
print([s for s in graph.migrate_vectors() if "place" in s])

['nodes.vec_place']


`source="city"` is the **property** holding the text to embed. It defaults to
the field's own name, so `Vector("city", 2, embed=…)` would have needed nothing
— the field name and the property it reads are the same word unless you say
otherwise.

With an embedder declared, a `set_vectors()` value may be a string:

In [23]:
graph.set_vectors(nodes=[{"id": ids["Alice"], "place": "Berlin"}])
print("Alice:", graph.get_vectors(node_ids=[ids["Alice"]])["nodes"][str(ids["Alice"])]["place"])

# Every string in the call is embedded in ONE request per field -- not one per
# row -- and all of it happens BEFORE the transaction opens. An HTTP call inside
# an open transaction would hold row locks for a network round trip, and a
# provider dying halfway would leave half a batch written for a retry to hit.
print(graph.set_vectors(nodes=[{"id": ids["Bob"], "place": "Berlin"},
                               {"id": ids["Carol"], "place": "Lisbon"}]), "rows")

Alice: [1.0, 0.0]
2 rows


And the query side: `Near(field, text=…)` embeds with the same field's client,
on the **query** side of the asymmetry — several providers score stored text and
query text differently, and getting that wrong raises nothing while quietly
costing recall. `embed_documents` and `embed_query` are that split, and hopai
picks the right one for you.

In [24]:
for hit in graph.vector_search(Near("place", text="Berlin"), k=3):
    print(f'{hit["properties"]["name"]:6} {hit["similarity"]:.2f}')

Alice  1.00
Bob    1.00
Carol  0.00


Alice and Bob at 1.00, Carol at 0.00 — exactly what `city_model` says, which is
the point of using a model you can read.

### The backfill

`embed_stale()` closes the loop: it takes every row `stale_vectors()` reports,
reads its `source` property, embeds them per field in one batched call, and
writes them. Rows whose property is missing or blank come back under `skipped`
rather than raising — Erin has no `city` key at all, and the two companies have
none either, so there is nothing to embed and nothing wrong:

In [25]:
print(graph.embed_stale(node_fields=["place"])["nodes"]["place"])
print()
for hit in graph.vector_search(Near("place", text="Lisbon"), k=2):
    print(f'{hit["properties"]["name"]:6} {hit["similarity"]:.2f}')

{'embedded': ['4'], 'skipped': ['5', '6', '7']}

Carol  1.00
Dave   1.00


Dave joins Carol in Lisbon, and the second run has nothing left to do.

One call finishes the field however large it is: it walks in pages of `batch`
rows, each its own embed call and its own transaction, so memory stays bounded
and a run that dies partway resumes rather than restarting. Re-running is for
picking up after a failure, not for making progress.

The paging is a cursor rather than a `LIMIT` window, and that is not a detail:
Erin and the two companies can never be filled in, so they are stale forever. A
window would hand back those same three rows on every pass and never reach the
work behind them — reporting success while embedding nothing.

Fields that declare no `embed=` are simply not this call's business: `summary`
and `title` are filled by hand and do not appear in the result at all.

## The two refusals

**Text yes, floats no.** A model may say what it is looking for — that is the
`text` above, embedded by the field with *your* client, so the query embedding
comes from the same model that wrote the stored ones. What it may never supply
is a `"vector"`: asked for floats a model invents plausible ones, and an
invented embedding finds confidently *wrong* neighbours with full confidence
scores attached. So `"vector"` is the one key the tool schemas never advertise:

In [26]:
import json

from hopai import (
    TRAVERSE_TOOL_SCHEMA, VECTOR_SEARCH_TOOL_SCHEMA, traverse_json, vector_search_json,
)

near_spec = TRAVERSE_TOOL_SCHEMA["parameters"]["$defs"]["near"]["properties"]
print("start:    ", sorted(TRAVERSE_TOOL_SCHEMA["parameters"]["properties"]["start"]["properties"]))
print("near spec:", sorted(near_spec))
print("search:   ", VECTOR_SEARCH_TOOL_SCHEMA["name"])
assert "text" in near_spec and "vector" not in near_spec

start:     ['boost', 'keep', 'near', 'where']
near spec: ['field', 'min_similarity', 'missing', 'text', 'weight']
search:    search_graph_by_meaning


The omission alone would be a convention, though, and a convention is not an
invariant. The JSON `"vector"` form exists for your own code — so the parsers
**refuse** it unless you say the spec did not come from a model, while the same
spec written with `"text"` runs untouched:

In [27]:
spec = {"start": {"near": {"field": "summary", "vector": databases}, "keep": 2}}

try:
    traverse_json(graph, spec)
except ValueError as exc:
    print(f"ValueError: {exc}\n")

allowed = traverse_json(graph, spec, allow_vectors=True)     # my own code, not a model
print("allow_vectors=True:", sorted(n["properties"]["name"] for n in allowed["nodes"]))

# What a tool call actually sends -- no floats anywhere, no opt-in needed.
from_a_model = traverse_json(graph, {
    "start": {"near": {"field": "place", "text": "Berlin"}, "keep": 2}})
print("from a model:      ", sorted(n["properties"]["name"] for n in from_a_model["nodes"]))
print("vector_search_json:", json.dumps(vector_search_json(
    graph, {"near": {"field": "place", "text": "Lisbon"}, "k": 2}))[:96], "...")

ValueError: traverse_json(): near={..."vector": [...]} cannot come from a tool call -- a model cannot supply a real embedding, and an invented one finds confidently wrong neighbors. Use {"field": 'summary', "text": "..."} and let the field embed it, or pass allow_vectors=True if this spec came from your own code



allow_vectors=True: ['Alice', 'Erin']
from a model:       ['Alice', 'Bob']
vector_search_json: {"results": [{"id": "3", "properties": {"age": 29, "city": "Lisbon", "name": "Carol", "type": "p ...


**A `Near` is not a filter.** It ranks; `where=` and `via=` take booleans. The
mistake is easy to make because `GT`/`BETWEEN` live exactly where a `Near`
looks like it should go, so both spellings are refused by name — with the
offending key named, so a five-key filter does not send you re-reading it:

In [28]:
for bad in (Near("summary", databases), {"summary": Near("summary", databases)}):
    try:
        graph.traverse(Start(where=bad))
    except TypeError as exc:
        print(f"TypeError: {exc}\n")

TypeError: Near(...) ranks rows by similarity; it is not a boolean filter. Pass it as near= on Start/Hop or to vector_search(), not inside where=/via=

TypeError: Near(...) ranks rows by similarity; it is not a boolean filter (in where={'summary': ...}). Pass it as near= on Start/Hop or to vector_search(), not inside where=/via=



## Re-embedding, and the door out

`embed_stale()` above is the automatic half. `stale_vectors()` is the report
under it, and the one you loop over yourself for fields you fill in by hand: the
rows that never had a vector, and those whose stored vector no longer matches
the declared dimensions — the window a dimension change opens, since
`migrate_vectors()` refuses to reinterpret stored vectors and `set_vectors()`
refuses to write the wrong size. Ids come back as strings, ready to hand
straight back:

In [29]:
stale = graph.stale_vectors()
print(stale["nodes"]["summary"], "<- Acme and Globex, which have no summary")
print(stale["nodes"]["title"], "  <- Erin too")

for node_id in stale["nodes"]["title"]["missing"][:1]:
    graph.set_vectors(nodes=[{"id": node_id, "title": [0.5, 0.5, 0.5]}])
print("after embedding one:", graph.stale_vectors(node_fields=["title"])["nodes"]["title"])

{'missing': ['6', '7'], 'wrong_dimensions': []} <- Acme and Globex, which have no summary
{'missing': ['5', '6', '7'], 'wrong_dimensions': []}   <- Erin too
after embedding one: {'missing': ['6', '7'], 'wrong_dimensions': []}


And if the exact scan is outgrown, that is a documented door rather than a
rewrite. `pgvector_exit_ddl()` prints the migration onto pgvector — generated
without importing, requiring or even checking for the extension:

In [30]:
print("\n".join(graph.pgvector_exit_ddl()[:4]))

CREATE EXTENSION IF NOT EXISTS vector
ALTER TABLE "nodes" DROP CONSTRAINT IF EXISTS "ck_vec_dims_default_nodes_summary"
ALTER TABLE "nodes" ALTER COLUMN "vec_summary" TYPE vector(4) USING "vec_summary"::vector(4)
CREATE INDEX IF NOT EXISTS "ix_nodes_vec_summary_hnsw" ON "nodes" USING hnsw ("vec_summary" vector_cosine_ops)


Read `hopai/vectors.py` before running it: the conversion is one-way, it makes
the search approximate, and the column is shared by every graph in the table.

## The SQL

The cosine is an `unnest` + `sum` in a **LATERAL**, one per field. That is a
measurement rather than a preference: written as a correlated scalar subquery
— which reads tidier — the planner pulled it up and re-evaluated it at every
site the outer query named it (the filter, the score, the `ORDER BY`), running
the `unnest` two to three times per candidate for identical results. It cost
2×; `benchmarks/README.md` has the numbers.

In [31]:
from sqlalchemy.dialects import postgresql

from hopai import Graph

offline = Graph("postgresql+psycopg2://nobody:nobody@127.0.0.1:1/nothing")   # never connects
offline.define_vectors(nodes=[Vector("summary", 4)])

print(offline.build_vector_search_query(Near("summary", databases), k=5)
             .compile(dialect=postgresql.dialect(),
                      compile_kwargs={"literal_binds": True}))

SELECT CAST(anon_1._id AS VARCHAR) AS id, anon_1.properties, anon_1.sim_0 AS similarity 
FROM (SELECT nodes.id AS _id, nodes.properties AS properties, near_0.s AS sim_0 
FROM nodes JOIN LATERAL (SELECT CASE WHEN (array_length(nodes.vec_summary, 1) = 4) THEN sum(CAST(anon_2.x AS DOUBLE PRECISION) * CAST(anon_2.y AS DOUBLE PRECISION)) / CAST(nullif(sqrt(sum(CAST(anon_2.x AS DOUBLE PRECISION) * CAST(anon_2.x AS DOUBLE PRECISION))) * 1.0, 0.0) AS DOUBLE PRECISION) END AS s 
FROM unnest(nodes.vec_summary, ARRAY[1.0, 0.0, 0.0, 0.0]) AS anon_2(x, y)) AS near_0 ON true 
WHERE nodes.graph_id = 'default' AND true AND nodes.vec_summary IS NOT NULL) AS anon_1 
WHERE anon_1.sim_0 IS NOT NULL ORDER BY similarity DESC, anon_1._id 
 LIMIT 5


Three details in that statement are load-bearing:

- **`JOIN LATERAL`**, as above.
- **`CASE WHEN (array_length(nodes.vec_summary, 1) = 4)`** — the wrong-length
  guard, which is what makes a mis-sized vector NULL instead of a confident
  cosine over whatever prefix the two arrays share.
- **`CAST(... AS DOUBLE PRECISION)` around each factor** — so the sum
  accumulates in float8. `sum(real)` accumulates in float4 and drifts
  measurably over embedding-length arrays.

And a traversal with no `near=` compiles to **byte-identical SQL** to the
pre-vector engine. A test asserts it, because "the vector feature slowed
everything else down a little" is exactly the kind of regression that hides.

## What this costs

The search is an exact scan, so its cost is linear in `candidates × dimensions`
— candidates meaning what is left *after* `where=`, which is why filtering is
the lever. The measured constant is **0.13 µs per vector element** on one
Postgres 16 core: an unfiltered 20 000 × 384-dim search takes about a second,
and the same search filtered to a quarter of the rows about 0.25 s.

So a few thousand filtered candidates is interactive, and an unfiltered scan
of hundreds of thousands is the wrong tool — at which point the exit door
above is the answer, not a workaround. `benchmarks/bench_vectors.py` measures
all of this rather than asserting it.

---

Back to [01 · Quick start](01_quickstart.ipynb), or on to `hopai/vectors.py`,
which argues every refusal in this notebook at length.